design/application.go

```diff
@@ -0,0 +1,174 @@
+		})
+
+	})
+
+	Method("register", func() {
+		Description("Register an Application with the FuseML application store.")
+
+		Payload(Application)
+
+		Error("BadRequest", func() {
+			Description("If the Application does not have the required fields, should return 400 Bad Request.")
+		})
+
+		Result(Application)
+
+		HTTP(func() {
+			POST("/applications")
+			Response(StatusCreated)
+			Response("BadRequest", StatusBadRequest)
+		})
+		GRPC(func() {
+			Response(CodeOK)
+			Response("BadRequest", CodeInvalidArgument)
+		})
+	})
+
+	Method("get", func() {
+		Description("Retrieve an Application registered by FuseML.")
+
+		Payload(func() {
+			Field(1, "name", String, "Application name", func() {
+				Example("mlflow-seldon-predictor-01")
+			})
+			Required("name")
+		})
+
+		Error("BadRequest", func() {
+			Description("If name is not given, should return 400 Bad Request.")
+		})
+		Error("NotFound", func() {
+			Description("If there is no application with the given name, should return 404 Not Found.")
+		})
+
+		Result(Application)
+
+		HTTP(func() {
+			GET("/applications/{name}")
+			Response(StatusOK)
+			Response("BadRequest", StatusBadRequest)
+			Response("NotFound", StatusNotFound)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("BadRequest", CodeInvalidArgument)
+			Response("NotFound", CodeNotFound)
+		})
+	})
+
+	Method("delete", func() {
+		Description("Delete an Application registered by FuseML application store.")
+
+		Payload(func() {
+			Field(1, "name", String, "Application name", func() {
+				Example("mlflow-seldon-predictor-01")
+			})
+			Field(2, "with_codeset", Boolean, "Delete related Codeset too", func() {
+				Default(false)
+			})
+			Required("name")
+		})
+
+		Error("BadRequest", func() {
+			Description("If the Application does not have the required fields, should return 400 Bad Request.")
+		})
+
+		HTTP(func() {
+			DELETE("/applications/{name}/")
+			Param("with_codeset", Boolean, "Delete related Codeset too", func() {
+				Default(false)
+			})
+			Response(StatusCreated)
+			Response("BadRequest", StatusBadRequest)
+		})
+		GRPC(func() {
+			Response(CodeOK)
+			Response("BadRequest", CodeInvalidArgument)
+		})
+	})
+
+})
+
+// Application describes the Application
+var Application = Type("Application", func() {
+
+	Field(1, "name", String, "The name of the Application", func() {
+		Example("mlflow-seldon-predictor-01")
+	})
+	Field(2, "type", String, "The type of the Application", func() {
+		Example("predictor")
+	})
+	Field(3, "description", String, "Application description", func() {
+		Example("My MLFlow predictor deployed via Seldon")
+	})
+	Field(4, "url", String, "The public URL for accessing the Application", func() {
+		Example("http://fuseml.example.org/mlflow-seldon-predictor-01/predict")
+	})
+	Field(5, "codeset", AppCodeset, "Codeset associated with the Application")
+	Field(6, "workflow", String, "Name of the Workflow used to create Application")
```

Comment:
Hm, I'm not so sure about it ...
You are right that one workflow can have several runs (based on some different input parameters for example in the codeset) which would produce different applications.

I'm just trying to imagine user's perspective. Users create a workflow (which I really think should rather be called recipe) and expect it to produce a result; having them dig into list of "workflow runs" seems like a PITA...

Reply:
Ha! You're adding yet another level of complexity on top of the level that I felt is too much! :-)

Here's the counter-proposal: there are no applications versions, application is _replaced_ with each new workflow run, thus application is really tied to the original workflow. Do you think it's a good idea to have new version of an application after each tiny change of data scientist code pushed into git repository?

Reply:
I guess it depends on what the user actually wants. The default behaviour should indeed be to replace the application with each new workflow run, because this is aligned with the gitops philosophy. We can explore alternative behaviours some other time and if we appreciate there's a need for them, we'll consider implementing them. For now, this is sufficient.